rockcraft/oci.py

```diff
@@ -156,15 +156,62 @@ def digest(self) -> bytes:
         parts = output.split(":", 1)
         return bytes.fromhex(parts[-1])
 
+    def set_entrypoint(self, entrypoint: List[str]) -> None:
+        """Set the OCI image entrypoint.
+
+        :param entrypoint: The default list of arguments to use as the command to
+            execute when the container starts.
+        """
+        emit.progress("Configuring entrypoint...")
+        image_path = self.path / self.image_name
+        params = ["--clear=config.entrypoint"]
+        for entry in entrypoint:
+            params.extend(["--config.entrypoint", entry])
+        _config_image(image_path, params)
+        emit.message(f"Entrypoint set to {entrypoint}", intermediate=True)
+
+    def set_cmd(self, cmd: List[str]) -> None:
+        """Set the OCI image default command parameters.
+
+        :param cmd: The default arguments to the entrypoint of the container. If
+            entrypoint is not defined, the first entry of cmd is the executable to
+            run when the container starts.
+        """
+        emit.progress("Configuring cmd...")
+        image_path = self.path / self.image_name
+        params = ["--clear=config.cmd"]
+        for entry in cmd:
+            params.extend(["--config.cmd", entry])
+        _config_image(image_path, params)
+        emit.message(f"Cmd set to {cmd}", intermediate=True)
+
+    def set_env(self, env: List[str]) -> None:
+        """Set the OCI image environment.
+
+        :param env: A list of environment variables in ``NAME=VALUE`` format.
+        """
+        emit.progress("Configuring env...")
+        image_path = self.path / self.image_name
+        params = ["--clear=config.env"]
+        for entry in env:
+            params.extend(["--config.env", entry])
+        _config_image(image_path, params)
+        emit.message(f"Environment set to {env}", intermediate=True)
+
 
 def _copy_image(source: str, destination: str) -> None:
     """Transfer images from source to destination."""
     _process_run(["skopeo", "--insecure-policy", "copy", source, destination])
 
 
-# XXX: update to use craft-cli output stream
+def _config_image(image_path: Path, params: List[str]) -> None:
+    """Configure the OCI image."""
+    _process_run(["umoci", "config", "--image", str(image_path)] + params)
```

Comment:
`umoci` creates new image config and manifest files, leaving the previous ones orphaned, but inside the same `blobs/sha256` directory. Are these being cleanup somewhere else?

Reply:
All the intermediate blobs exist inside the internal working directories, but when the final artifact (the `.rock` file) is exported it only contains the blobs corresponding to the final tag. The working directories are created inside the build container and are not exposed to the user.

Working directory content:
```
 root@rockcraft-hello-1987971:~# ls -l images/hello/blobs/sha256/
total 28168
-rw-r--r-- 1 root root      349 Apr 14 12:42 095ee187a14c5b2f970d09c1c31e3583791aba5290f78dbf587f6bbe6e3405a0
-rw-r--r-- 1 root root      577 Apr 14 12:42 14814774e98aa9a8be826f1155fc6349e2b750809513eeb5a2aeb98c55a88a33
-rw------- 1 root root      732 Apr 14 12:43 21c1b6182858be8dcc70ef3566ee80082147a7ef5224e38d3e78a435fec47c88
-rw------- 1 root root      864 Apr 14 12:43 2f70e21812fc84eb308586b9da197ce3b7dc0074ea460d29f28fc0e3fab54fc1
-rw------- 1 root root      507 Apr 14 12:43 4b14b089a90aa18e15a9cd1d0ce999d858169996d0eb38d7a57e92b25f03e49a
-rw------- 1 root root   239622 Apr 14 12:43 549607a70eeea9c42d2687ef8ababef8d36a5157727b016a364d220ac196ea47
-rw------- 1 root root      507 Apr 14 12:43 814d03fb0110de8d5fc618b0fc55ff7222e6fbe71290556436b9dd22efd245e8
-rw------- 1 root root      507 Apr 14 12:43 8f380bcd899f35a2d885a0639c3dea81f1695da965866e837e66387ecdcdc926
-rw-r--r-- 1 root root 28566292 Apr 14 12:42 e0b25ef516347a097d75f8aea6bc0f42a4e8e70b057e84d85098d51f96d458f9
-rw------- 1 root root      828 Apr 14 12:43 e177c7ea890f8a6eb35b4c9cc1d2491cb7b4f029ec73431ac7131f739254d84b
```

Final artifact:
```
$ tar tvf hello_1.0.rock
drwxr-xr-x root/root         0 2022-04-14 09:43 blobs/
drwxr-xr-x root/root         0 2022-04-14 09:43 blobs/sha256/
-rw-r--r-- root/root       864 2022-04-14 09:43 blobs/sha256/2f70e21812fc84eb308586b9da197ce3b7dc0074ea460d29f28fc0e3fab54fc1
-rw-r--r-- root/root    239622 2022-04-14 09:43 blobs/sha256/549607a70eeea9c42d2687ef8ababef8d36a5157727b016a364d220ac196ea47
-rw-r--r-- root/root       507 2022-04-14 09:43 blobs/sha256/814d03fb0110de8d5fc618b0fc55ff7222e6fbe71290556436b9dd22efd245e8
-rw-r--r-- root/root  28566292 2022-04-14 09:43 blobs/sha256/e0b25ef516347a097d75f8aea6bc0f42a4e8e70b057e84d85098d51f96d458f9
-rw-r--r-- root/root       244 2022-04-14 09:43 index.json
-rw-r--r-- root/root        31 2022-04-14 09:43 oci-layout
```